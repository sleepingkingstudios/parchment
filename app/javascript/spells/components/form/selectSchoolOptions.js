const selectSchoolsOptions = [
  {
    label: 'Abjuration',
    value: 'abjuration',
  },
  {
    label: 'Conjuration',
    value: 'conjuration',
  },
  {
    label: 'Divination',
    value: 'divination',
  },
  {
    label: 'Enchantment',
    value: 'enchantment',
  },
  {
    label: 'Evocation',
    value: 'evocation',
  },
  {
    label: 'Illusion',
    value: 'illusion',
  },
  {
    label: 'Necromancy',
    value: 'necromancy',
  },
  {
    label: 'Transmutation',
    value: 'transmutation',
  },
];

export default selectSchoolsOptions;
