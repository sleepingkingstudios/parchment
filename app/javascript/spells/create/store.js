import submitForm from '../../requests/submitForm';
import { buildSpell } from '../entities';

const namespace = 'createSpell';
const REQUEST_URL = '/api/spells';

export const {
  actions,
  reducer,
  request,
} = submitForm({
  data: { spell: buildSpell() },
  namespace,
  url: REQUEST_URL,
});
