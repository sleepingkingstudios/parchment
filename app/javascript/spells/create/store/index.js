import submitForm from '../../../requests/submitForm';
import { buildSpell } from '../../entities';
import alerts from './alerts';
import redirect from './redirect';

const namespace = 'createSpell';
const REQUEST_URL = '/api/spells';

export const {
  actions,
  reducer,
  request,
} = submitForm({
  data: { spell: buildSpell() },
  middleware: [
    redirect,
    alerts,
  ],
  namespace,
  url: REQUEST_URL,
});
