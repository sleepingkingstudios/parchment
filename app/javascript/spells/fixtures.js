/* eslint-disable-next-line import/prefer-default-export */
export const spellsData = [
  {
    id: 'e17bca01-24d3-48d3-a308-099dd6dca373',
    level: 0,
    name: 'Embers',
    school: 'Evocation',
  },
  {
    id: 'e3641970-fc98-4674-b873-5532d89d9f26',
    level: 3,
    name: "Dragon's Breath",
    school: 'Evocation',
  },
  {
    id: 'a74d8c21-607d-4a3f-8cfd-2bbb48ec196a',
    level: 9,
    name: 'Wrath of the Titans',
    school: 'Evocation',
  },
];
