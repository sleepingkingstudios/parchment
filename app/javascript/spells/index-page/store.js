import buildStore from 'resource/pages/index-page/store';

const namespace = 'spells/index';
const resourceName = 'spells';
const url = 'api/spells';

const store = buildStore({
  namespace,
  resourceName,
  url,
});
const {
  hooks,
  reducer,
} = store;

export default store;

export {
  hooks,
  reducer,
};
