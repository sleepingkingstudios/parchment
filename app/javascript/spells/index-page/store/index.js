import { combineReducers } from 'redux';

import { hooks as dataHooks, reducer as data } from './data';
import { hooks as destroyHooks, reducer as destroy } from './destroy';
import { hooks as findHooks, reducer as find } from './find';

const { useData } = dataHooks;
const useDataStatus = findHooks.useStatus;
const useRequestData = findHooks.usePerformRequest;
const useDestroyRequest = destroyHooks.useRequest;

export const hooks = {
  useData,
  useDataStatus,
  useDestroyRequest,
  useRequestData,
};

export const reducer = combineReducers({
  data,
  destroy,
  find,
});
