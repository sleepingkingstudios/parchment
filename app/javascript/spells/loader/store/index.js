import alerts from './alerts';
import redirect from './redirect';
import findOne from '../../../requests/findOne';
import { buildSpell } from '../../entities';

const namespace = 'findSpell';
const REQUEST_URL = '/api/spells/:id';

export const {
  actions,
  reducer,
  request,
} = findOne({
  data: { spell: buildSpell() },
  middleware: [
    redirect,
    alerts,
  ],
  namespace,
  url: REQUEST_URL,
});
