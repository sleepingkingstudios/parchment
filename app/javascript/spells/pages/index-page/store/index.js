import { combineReducers } from 'redux';

import { hooks as findHooks, reducer as find } from './find';

const useRequestData = findHooks.usePerformRequest;

export const hooks = {
  useRequestData,
};

export const reducer = combineReducers({
  find,
});
