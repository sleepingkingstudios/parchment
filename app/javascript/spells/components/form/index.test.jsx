import React from 'react';
import { shallow } from 'enzyme';

import SpellForm from './index';

import { spellsData } from '../../fixtures';
import selectSchoolOptions from './selectSchoolOptions';
import { INITIALIZED } from '../../../api/status';
import { hooks } from '../../store/formFindOrigins';

jest.mock('../../store/formFindOrigins');

hooks.useRequestData.mockImplementation(() => () => {});

describe('<SpellForm />', () => {
  const onChangeAction = jest.fn(
    ({ propName, value }) => ({ payload: { propName, value } }),
  );
  const onSubmitAction = jest.fn(() => ({ ok: true }));
  const data = { spell: spellsData[0] };
  const errors = {};
  const defaultProps = {
    data,
    onChangeAction,
    onSubmitAction,
    status: INITIALIZED,
    errors,
  };
  const form = {
    data,
    errors,
    path: ['spell'],
    onChangeAction,
    onSubmitAction,
  };

  it('should render a form', () => {
    const rendered = shallow(<SpellForm {...defaultProps} />);

    expect(rendered).toHaveDisplayName('Form');
    expect(rendered).toHaveProp({ form });
  });

  it('should render the name input field', () => {
    const rendered = shallow(<SpellForm {...defaultProps} />);
    const input = rendered.find('NameField');

    expect(input).toExist();
    expect(input).toHaveProp({ form });
  });

  it('should render the slug field', () => {
    const rendered = shallow(<SpellForm {...defaultProps} />);
    const input = rendered.find('SlugField');

    expect(input).toExist();
    expect(input).toHaveProp({ form });
  });

  it('should render the school select field', () => {
    const rendered = shallow(<SpellForm {...defaultProps} />);
    const input = rendered.find('SchoolField');
    const options = selectSchoolOptions;

    expect(input).toExist();
    expect(input).toHaveProp({ form });
    expect(input).toHaveProp({ options });
  });

  it('should render the level input field', () => {
    const rendered = shallow(<SpellForm {...defaultProps} />);
    const input = rendered.find('LevelField');

    expect(input).toExist();
    expect(input).toHaveProp({ form });
  });

  it('should render the ritual checkbox field', () => {
    const rendered = shallow(<SpellForm {...defaultProps} />);
    const input = rendered.find('RitualField');

    expect(input).toExist();
    expect(input).toHaveProp({ form });
  });

  it('should render the source select field', () => {
    const rendered = shallow(<SpellForm {...defaultProps} />);
    const input = rendered.find('SourceField');

    expect(input).toExist();
    expect(input).toHaveProp({ form });
  });

  it('should render the casting time input field', () => {
    const rendered = shallow(<SpellForm {...defaultProps} />);
    const input = rendered.find('CastingTimeField');

    expect(input).toExist();
    expect(input).toHaveProp({ form });
  });

  it('should render the range input field', () => {
    const rendered = shallow(<SpellForm {...defaultProps} />);
    const input = rendered.find('RangeField');

    expect(input).toExist();
    expect(input).toHaveProp({ form });
  });

  it('should render the duration input field', () => {
    const rendered = shallow(<SpellForm {...defaultProps} />);
    const input = rendered.find('DurationField');

    expect(input).toExist();
    expect(input).toHaveProp({ form });
  });

  it('should render the components field', () => {
    const rendered = shallow(<SpellForm {...defaultProps} />);
    const input = rendered.find('SpellFormComponentsField');

    expect(input).toExist();
    expect(input).toHaveProp({ form });
  });

  it('should render the short description field', () => {
    const rendered = shallow(<SpellForm {...defaultProps} />);
    const input = rendered.find('ShortDescriptionField');

    expect(input).toExist();
    expect(input).toHaveProp({ form });
  });

  it('should render the description textarea field', () => {
    const rendered = shallow(<SpellForm {...defaultProps} />);
    const input = rendered.find('DescriptionField');

    expect(input).toExist();
    expect(input).toHaveProp({ form });
  });

  it('should render the cancel button', () => {
    const rendered = shallow(<SpellForm {...defaultProps} />);
    const input = rendered.find('CancelButton');

    expect(input).toExist();
    expect(input).toHaveProp({ form });
  });

  it('should render the submit button', () => {
    const rendered = shallow(<SpellForm {...defaultProps} />);
    const input = rendered.find('SubmitButton');

    expect(input).toExist();
    expect(input).toHaveProp({ form });
  });

  it('should find the origins', () => {
    const performRequest = jest.fn();

    hooks.useRequestData.mockImplementationOnce(() => performRequest);

    shallow(<SpellForm {...defaultProps} />);

    expect(hooks.useRequestData).toHaveBeenCalled();
    expect(performRequest).toHaveBeenCalled();
  });

  it('should match the snapshot', () => {
    const rendered = shallow(<SpellForm {...defaultProps} />);

    expect(rendered).toMatchSnapshot();
  });
});
