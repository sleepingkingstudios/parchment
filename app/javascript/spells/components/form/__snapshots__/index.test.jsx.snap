// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SpellForm /> should match the snapshot 1`] = `
<Form
  className="spell-form"
  form={
    Object {
      "data": Object {
        "spell": Object {
          "castingTime": "1 action",
          "description": "You conjure a glowing, spectral orb the size of a closed fist and bearing the likeness of everyone's favorite creature, the Flumph. Until the spell ends, the orb sheds bright light in a 20-foot radius and dim light for an additional 20 feet. The light can be colored as you like. The spell ends if you cast it again or dismiss it as an action.

You can move the orb to anywhere within 30 feet of you as a bonus action. If you move more than 30 feet away from the orb, it will float directly to your location at the beginning of your next turn; if it is unable to reach your location, the spell immediately ends. ",
          "duration": "1 hour",
          "id": "823dbe98-ef01-4984-b8b2-34c2946091ce",
          "level": 0,
          "materialComponent": "a pinch of flumph dust",
          "name": "Flumph Lantern",
          "range": "30 feet",
          "ritual": false,
          "school": "evocation",
          "shortDescription": "Floating light illuminates 20 feet + 20 feet dim light",
          "slug": "flumph-lantern",
          "somaticComponent": false,
          "sourceId": "da08708e-0505-4819-bc69-00b6589a4e4f",
          "sourceType": "Publication",
          "verbalComponent": true,
        },
      },
      "errors": Object {},
      "onChangeAction": [MockFunction],
      "onSubmitAction": [MockFunction],
      "path": Array [
        "spell",
      ],
    }
  }
>
  <FormRow
    align="left"
    className=""
    verticalAlign="top"
  >
    <NameField
      colWidth="8"
      form={
        Object {
          "data": Object {
            "spell": Object {
              "castingTime": "1 action",
              "description": "You conjure a glowing, spectral orb the size of a closed fist and bearing the likeness of everyone's favorite creature, the Flumph. Until the spell ends, the orb sheds bright light in a 20-foot radius and dim light for an additional 20 feet. The light can be colored as you like. The spell ends if you cast it again or dismiss it as an action.

You can move the orb to anywhere within 30 feet of you as a bonus action. If you move more than 30 feet away from the orb, it will float directly to your location at the beginning of your next turn; if it is unable to reach your location, the spell immediately ends. ",
              "duration": "1 hour",
              "id": "823dbe98-ef01-4984-b8b2-34c2946091ce",
              "level": 0,
              "materialComponent": "a pinch of flumph dust",
              "name": "Flumph Lantern",
              "range": "30 feet",
              "ritual": false,
              "school": "evocation",
              "shortDescription": "Floating light illuminates 20 feet + 20 feet dim light",
              "slug": "flumph-lantern",
              "somaticComponent": false,
              "sourceId": "da08708e-0505-4819-bc69-00b6589a4e4f",
              "sourceType": "Publication",
              "verbalComponent": true,
            },
          },
          "errors": Object {},
          "onChangeAction": [MockFunction],
          "onSubmitAction": [MockFunction],
          "path": Array [
            "spell",
          ],
        }
      }
    />
    <SlugField
      colWidth="4"
      form={
        Object {
          "data": Object {
            "spell": Object {
              "castingTime": "1 action",
              "description": "You conjure a glowing, spectral orb the size of a closed fist and bearing the likeness of everyone's favorite creature, the Flumph. Until the spell ends, the orb sheds bright light in a 20-foot radius and dim light for an additional 20 feet. The light can be colored as you like. The spell ends if you cast it again or dismiss it as an action.

You can move the orb to anywhere within 30 feet of you as a bonus action. If you move more than 30 feet away from the orb, it will float directly to your location at the beginning of your next turn; if it is unable to reach your location, the spell immediately ends. ",
              "duration": "1 hour",
              "id": "823dbe98-ef01-4984-b8b2-34c2946091ce",
              "level": 0,
              "materialComponent": "a pinch of flumph dust",
              "name": "Flumph Lantern",
              "range": "30 feet",
              "ritual": false,
              "school": "evocation",
              "shortDescription": "Floating light illuminates 20 feet + 20 feet dim light",
              "slug": "flumph-lantern",
              "somaticComponent": false,
              "sourceId": "da08708e-0505-4819-bc69-00b6589a4e4f",
              "sourceType": "Publication",
              "verbalComponent": true,
            },
          },
          "errors": Object {},
          "onChangeAction": [MockFunction],
          "onSubmitAction": [MockFunction],
          "path": Array [
            "spell",
          ],
        }
      }
    />
    <SchoolField
      colWidth="8"
      form={
        Object {
          "data": Object {
            "spell": Object {
              "castingTime": "1 action",
              "description": "You conjure a glowing, spectral orb the size of a closed fist and bearing the likeness of everyone's favorite creature, the Flumph. Until the spell ends, the orb sheds bright light in a 20-foot radius and dim light for an additional 20 feet. The light can be colored as you like. The spell ends if you cast it again or dismiss it as an action.

You can move the orb to anywhere within 30 feet of you as a bonus action. If you move more than 30 feet away from the orb, it will float directly to your location at the beginning of your next turn; if it is unable to reach your location, the spell immediately ends. ",
              "duration": "1 hour",
              "id": "823dbe98-ef01-4984-b8b2-34c2946091ce",
              "level": 0,
              "materialComponent": "a pinch of flumph dust",
              "name": "Flumph Lantern",
              "range": "30 feet",
              "ritual": false,
              "school": "evocation",
              "shortDescription": "Floating light illuminates 20 feet + 20 feet dim light",
              "slug": "flumph-lantern",
              "somaticComponent": false,
              "sourceId": "da08708e-0505-4819-bc69-00b6589a4e4f",
              "sourceType": "Publication",
              "verbalComponent": true,
            },
          },
          "errors": Object {},
          "onChangeAction": [MockFunction],
          "onSubmitAction": [MockFunction],
          "path": Array [
            "spell",
          ],
        }
      }
      options={
        Array [
          Object {
            "label": "Abjuration",
            "value": "abjuration",
          },
          Object {
            "label": "Conjuration",
            "value": "conjuration",
          },
          Object {
            "label": "Divination",
            "value": "divination",
          },
          Object {
            "label": "Enchantment",
            "value": "enchantment",
          },
          Object {
            "label": "Evocation",
            "value": "evocation",
          },
          Object {
            "label": "Illusion",
            "value": "illusion",
          },
          Object {
            "label": "Necromancy",
            "value": "necromancy",
          },
          Object {
            "label": "Transmutation",
            "value": "transmutation",
          },
        ]
      }
    />
    <LevelField
      colWidth="2"
      form={
        Object {
          "data": Object {
            "spell": Object {
              "castingTime": "1 action",
              "description": "You conjure a glowing, spectral orb the size of a closed fist and bearing the likeness of everyone's favorite creature, the Flumph. Until the spell ends, the orb sheds bright light in a 20-foot radius and dim light for an additional 20 feet. The light can be colored as you like. The spell ends if you cast it again or dismiss it as an action.

You can move the orb to anywhere within 30 feet of you as a bonus action. If you move more than 30 feet away from the orb, it will float directly to your location at the beginning of your next turn; if it is unable to reach your location, the spell immediately ends. ",
              "duration": "1 hour",
              "id": "823dbe98-ef01-4984-b8b2-34c2946091ce",
              "level": 0,
              "materialComponent": "a pinch of flumph dust",
              "name": "Flumph Lantern",
              "range": "30 feet",
              "ritual": false,
              "school": "evocation",
              "shortDescription": "Floating light illuminates 20 feet + 20 feet dim light",
              "slug": "flumph-lantern",
              "somaticComponent": false,
              "sourceId": "da08708e-0505-4819-bc69-00b6589a4e4f",
              "sourceType": "Publication",
              "verbalComponent": true,
            },
          },
          "errors": Object {},
          "onChangeAction": [MockFunction],
          "onSubmitAction": [MockFunction],
          "path": Array [
            "spell",
          ],
        }
      }
    />
    <RitualField
      colWidth="2"
      form={
        Object {
          "data": Object {
            "spell": Object {
              "castingTime": "1 action",
              "description": "You conjure a glowing, spectral orb the size of a closed fist and bearing the likeness of everyone's favorite creature, the Flumph. Until the spell ends, the orb sheds bright light in a 20-foot radius and dim light for an additional 20 feet. The light can be colored as you like. The spell ends if you cast it again or dismiss it as an action.

You can move the orb to anywhere within 30 feet of you as a bonus action. If you move more than 30 feet away from the orb, it will float directly to your location at the beginning of your next turn; if it is unable to reach your location, the spell immediately ends. ",
              "duration": "1 hour",
              "id": "823dbe98-ef01-4984-b8b2-34c2946091ce",
              "level": 0,
              "materialComponent": "a pinch of flumph dust",
              "name": "Flumph Lantern",
              "range": "30 feet",
              "ritual": false,
              "school": "evocation",
              "shortDescription": "Floating light illuminates 20 feet + 20 feet dim light",
              "slug": "flumph-lantern",
              "somaticComponent": false,
              "sourceId": "da08708e-0505-4819-bc69-00b6589a4e4f",
              "sourceType": "Publication",
              "verbalComponent": true,
            },
          },
          "errors": Object {},
          "onChangeAction": [MockFunction],
          "onSubmitAction": [MockFunction],
          "path": Array [
            "spell",
          ],
        }
      }
    />
    <CastingTimeField
      colWidth="4"
      form={
        Object {
          "data": Object {
            "spell": Object {
              "castingTime": "1 action",
              "description": "You conjure a glowing, spectral orb the size of a closed fist and bearing the likeness of everyone's favorite creature, the Flumph. Until the spell ends, the orb sheds bright light in a 20-foot radius and dim light for an additional 20 feet. The light can be colored as you like. The spell ends if you cast it again or dismiss it as an action.

You can move the orb to anywhere within 30 feet of you as a bonus action. If you move more than 30 feet away from the orb, it will float directly to your location at the beginning of your next turn; if it is unable to reach your location, the spell immediately ends. ",
              "duration": "1 hour",
              "id": "823dbe98-ef01-4984-b8b2-34c2946091ce",
              "level": 0,
              "materialComponent": "a pinch of flumph dust",
              "name": "Flumph Lantern",
              "range": "30 feet",
              "ritual": false,
              "school": "evocation",
              "shortDescription": "Floating light illuminates 20 feet + 20 feet dim light",
              "slug": "flumph-lantern",
              "somaticComponent": false,
              "sourceId": "da08708e-0505-4819-bc69-00b6589a4e4f",
              "sourceType": "Publication",
              "verbalComponent": true,
            },
          },
          "errors": Object {},
          "onChangeAction": [MockFunction],
          "onSubmitAction": [MockFunction],
          "path": Array [
            "spell",
          ],
        }
      }
    />
    <RangeField
      colWidth="4"
      form={
        Object {
          "data": Object {
            "spell": Object {
              "castingTime": "1 action",
              "description": "You conjure a glowing, spectral orb the size of a closed fist and bearing the likeness of everyone's favorite creature, the Flumph. Until the spell ends, the orb sheds bright light in a 20-foot radius and dim light for an additional 20 feet. The light can be colored as you like. The spell ends if you cast it again or dismiss it as an action.

You can move the orb to anywhere within 30 feet of you as a bonus action. If you move more than 30 feet away from the orb, it will float directly to your location at the beginning of your next turn; if it is unable to reach your location, the spell immediately ends. ",
              "duration": "1 hour",
              "id": "823dbe98-ef01-4984-b8b2-34c2946091ce",
              "level": 0,
              "materialComponent": "a pinch of flumph dust",
              "name": "Flumph Lantern",
              "range": "30 feet",
              "ritual": false,
              "school": "evocation",
              "shortDescription": "Floating light illuminates 20 feet + 20 feet dim light",
              "slug": "flumph-lantern",
              "somaticComponent": false,
              "sourceId": "da08708e-0505-4819-bc69-00b6589a4e4f",
              "sourceType": "Publication",
              "verbalComponent": true,
            },
          },
          "errors": Object {},
          "onChangeAction": [MockFunction],
          "onSubmitAction": [MockFunction],
          "path": Array [
            "spell",
          ],
        }
      }
    />
    <DurationField
      colWidth="4"
      form={
        Object {
          "data": Object {
            "spell": Object {
              "castingTime": "1 action",
              "description": "You conjure a glowing, spectral orb the size of a closed fist and bearing the likeness of everyone's favorite creature, the Flumph. Until the spell ends, the orb sheds bright light in a 20-foot radius and dim light for an additional 20 feet. The light can be colored as you like. The spell ends if you cast it again or dismiss it as an action.

You can move the orb to anywhere within 30 feet of you as a bonus action. If you move more than 30 feet away from the orb, it will float directly to your location at the beginning of your next turn; if it is unable to reach your location, the spell immediately ends. ",
              "duration": "1 hour",
              "id": "823dbe98-ef01-4984-b8b2-34c2946091ce",
              "level": 0,
              "materialComponent": "a pinch of flumph dust",
              "name": "Flumph Lantern",
              "range": "30 feet",
              "ritual": false,
              "school": "evocation",
              "shortDescription": "Floating light illuminates 20 feet + 20 feet dim light",
              "slug": "flumph-lantern",
              "somaticComponent": false,
              "sourceId": "da08708e-0505-4819-bc69-00b6589a4e4f",
              "sourceType": "Publication",
              "verbalComponent": true,
            },
          },
          "errors": Object {},
          "onChangeAction": [MockFunction],
          "onSubmitAction": [MockFunction],
          "path": Array [
            "spell",
          ],
        }
      }
    />
    <SpellFormComponentsField
      colWidth="12"
      form={
        Object {
          "data": Object {
            "spell": Object {
              "castingTime": "1 action",
              "description": "You conjure a glowing, spectral orb the size of a closed fist and bearing the likeness of everyone's favorite creature, the Flumph. Until the spell ends, the orb sheds bright light in a 20-foot radius and dim light for an additional 20 feet. The light can be colored as you like. The spell ends if you cast it again or dismiss it as an action.

You can move the orb to anywhere within 30 feet of you as a bonus action. If you move more than 30 feet away from the orb, it will float directly to your location at the beginning of your next turn; if it is unable to reach your location, the spell immediately ends. ",
              "duration": "1 hour",
              "id": "823dbe98-ef01-4984-b8b2-34c2946091ce",
              "level": 0,
              "materialComponent": "a pinch of flumph dust",
              "name": "Flumph Lantern",
              "range": "30 feet",
              "ritual": false,
              "school": "evocation",
              "shortDescription": "Floating light illuminates 20 feet + 20 feet dim light",
              "slug": "flumph-lantern",
              "somaticComponent": false,
              "sourceId": "da08708e-0505-4819-bc69-00b6589a4e4f",
              "sourceType": "Publication",
              "verbalComponent": true,
            },
          },
          "errors": Object {},
          "onChangeAction": [MockFunction],
          "onSubmitAction": [MockFunction],
          "path": Array [
            "spell",
          ],
        }
      }
    />
    <ShortDescriptionField
      colWidth="12"
      form={
        Object {
          "data": Object {
            "spell": Object {
              "castingTime": "1 action",
              "description": "You conjure a glowing, spectral orb the size of a closed fist and bearing the likeness of everyone's favorite creature, the Flumph. Until the spell ends, the orb sheds bright light in a 20-foot radius and dim light for an additional 20 feet. The light can be colored as you like. The spell ends if you cast it again or dismiss it as an action.

You can move the orb to anywhere within 30 feet of you as a bonus action. If you move more than 30 feet away from the orb, it will float directly to your location at the beginning of your next turn; if it is unable to reach your location, the spell immediately ends. ",
              "duration": "1 hour",
              "id": "823dbe98-ef01-4984-b8b2-34c2946091ce",
              "level": 0,
              "materialComponent": "a pinch of flumph dust",
              "name": "Flumph Lantern",
              "range": "30 feet",
              "ritual": false,
              "school": "evocation",
              "shortDescription": "Floating light illuminates 20 feet + 20 feet dim light",
              "slug": "flumph-lantern",
              "somaticComponent": false,
              "sourceId": "da08708e-0505-4819-bc69-00b6589a4e4f",
              "sourceType": "Publication",
              "verbalComponent": true,
            },
          },
          "errors": Object {},
          "onChangeAction": [MockFunction],
          "onSubmitAction": [MockFunction],
          "path": Array [
            "spell",
          ],
        }
      }
    />
    <DescriptionField
      colWidth="12"
      form={
        Object {
          "data": Object {
            "spell": Object {
              "castingTime": "1 action",
              "description": "You conjure a glowing, spectral orb the size of a closed fist and bearing the likeness of everyone's favorite creature, the Flumph. Until the spell ends, the orb sheds bright light in a 20-foot radius and dim light for an additional 20 feet. The light can be colored as you like. The spell ends if you cast it again or dismiss it as an action.

You can move the orb to anywhere within 30 feet of you as a bonus action. If you move more than 30 feet away from the orb, it will float directly to your location at the beginning of your next turn; if it is unable to reach your location, the spell immediately ends. ",
              "duration": "1 hour",
              "id": "823dbe98-ef01-4984-b8b2-34c2946091ce",
              "level": 0,
              "materialComponent": "a pinch of flumph dust",
              "name": "Flumph Lantern",
              "range": "30 feet",
              "ritual": false,
              "school": "evocation",
              "shortDescription": "Floating light illuminates 20 feet + 20 feet dim light",
              "slug": "flumph-lantern",
              "somaticComponent": false,
              "sourceId": "da08708e-0505-4819-bc69-00b6589a4e4f",
              "sourceType": "Publication",
              "verbalComponent": true,
            },
          },
          "errors": Object {},
          "onChangeAction": [MockFunction],
          "onSubmitAction": [MockFunction],
          "path": Array [
            "spell",
          ],
        }
      }
      rows={5}
    />
  </FormRow>
  <FormRow
    align="right"
    className=""
    verticalAlign="top"
  >
    <CancelButton
      colWidth="3"
      form={
        Object {
          "data": Object {
            "spell": Object {
              "castingTime": "1 action",
              "description": "You conjure a glowing, spectral orb the size of a closed fist and bearing the likeness of everyone's favorite creature, the Flumph. Until the spell ends, the orb sheds bright light in a 20-foot radius and dim light for an additional 20 feet. The light can be colored as you like. The spell ends if you cast it again or dismiss it as an action.

You can move the orb to anywhere within 30 feet of you as a bonus action. If you move more than 30 feet away from the orb, it will float directly to your location at the beginning of your next turn; if it is unable to reach your location, the spell immediately ends. ",
              "duration": "1 hour",
              "id": "823dbe98-ef01-4984-b8b2-34c2946091ce",
              "level": 0,
              "materialComponent": "a pinch of flumph dust",
              "name": "Flumph Lantern",
              "range": "30 feet",
              "ritual": false,
              "school": "evocation",
              "shortDescription": "Floating light illuminates 20 feet + 20 feet dim light",
              "slug": "flumph-lantern",
              "somaticComponent": false,
              "sourceId": "da08708e-0505-4819-bc69-00b6589a4e4f",
              "sourceType": "Publication",
              "verbalComponent": true,
            },
          },
          "errors": Object {},
          "onChangeAction": [MockFunction],
          "onSubmitAction": [MockFunction],
          "path": Array [
            "spell",
          ],
        }
      }
      isUpdate={false}
    />
    <SubmitButton
      colWidth="3"
      disabled={false}
      form={
        Object {
          "data": Object {
            "spell": Object {
              "castingTime": "1 action",
              "description": "You conjure a glowing, spectral orb the size of a closed fist and bearing the likeness of everyone's favorite creature, the Flumph. Until the spell ends, the orb sheds bright light in a 20-foot radius and dim light for an additional 20 feet. The light can be colored as you like. The spell ends if you cast it again or dismiss it as an action.

You can move the orb to anywhere within 30 feet of you as a bonus action. If you move more than 30 feet away from the orb, it will float directly to your location at the beginning of your next turn; if it is unable to reach your location, the spell immediately ends. ",
              "duration": "1 hour",
              "id": "823dbe98-ef01-4984-b8b2-34c2946091ce",
              "level": 0,
              "materialComponent": "a pinch of flumph dust",
              "name": "Flumph Lantern",
              "range": "30 feet",
              "ritual": false,
              "school": "evocation",
              "shortDescription": "Floating light illuminates 20 feet + 20 feet dim light",
              "slug": "flumph-lantern",
              "somaticComponent": false,
              "sourceId": "da08708e-0505-4819-bc69-00b6589a4e4f",
              "sourceType": "Publication",
              "verbalComponent": true,
            },
          },
          "errors": Object {},
          "onChangeAction": [MockFunction],
          "onSubmitAction": [MockFunction],
          "path": Array [
            "spell",
          ],
        }
      }
      isUpdate={false}
      status="initialized"
    />
  </FormRow>
</Form>
`;
